README.md
# fluent-plugin-gcloud-pubsub-custom

[![Build Status](https://travis-ci.org/mia-0032/fluent-plugin-gcloud-pubsub-custom.svg?branch=master)](https://travis-ci.org/mia-0032/fluent-plugin-gcloud-pubsub-custom)
[![Gem Version](https://badge.fury.io/rb/fluent-plugin-gcloud-pubsub-custom.svg)](http://badge.fury.io/rb/fluent-plugin-gcloud-pubsub-custom)

This plugin is forked from https://github.com/mdoi/fluent-plugin-gcloud-pubsub

## Overview

[Google Cloud Pub/Sub](https://cloud.google.com/pubsub/) Input/Output(BufferedOutput) plugin for [Fluentd](http://www.fluentd.org/) with [google-cloud](https://googlecloudplatform.github.io/google-cloud-ruby/) gem

- Publish messages to Google Cloud Pub/Sub
- Pull messages from Google Cloud Pub/Sub

## Preparation

- Create a project on Google Developer Console
- Add a topic of Cloud Pub/Sub to the project
- Add a pull style subscription to the topic
- Download your credential (json) or [set scope on GCE instance](https://cloud.google.com/compute/docs/api/how-tos/authorization)

## Requirements

| fluent-plugin-gcloud-pubsub-custom | fluentd | ruby |
|------------------------|---------|------|
| >= 1.0.0 | >= v0.14.0 | >= 2.1 |
|  < 1.0.0 | >= v0.12.0 | >= 1.9 |

## Installation

Install by gem:

```shell
$ gem install fluent-plugin-gcloud-pubsub-custom
```

**Caution**

This plugin doesn't work in [td-agent](http://docs.fluentd.org/articles/install-by-rpm).

Please use in [Fluentd installed by gem](http://docs.fluentd.org/articles/install-by-gem).

## Configuration

### Publish messages

Use `gcloud_pubsub` output plugin.

```
<match example.publish>
  @type gcloud_pubsub
  project <YOUR PROJECT>
  key <YOUR KEY>
  topic <YOUR TOPIC>
  autocreate_topic false
  max_messages 1000
  max_total_size 9800000
  max_message_size 4000000
  buffer_type file
  buffer_path /path/to/your/buffer
  flush_interval 1s
  try_flush_interval 0.1
  format json
</match>
```

- `project` (optional)
  - Set your GCP project.
  - Running fluentd on GCP, you don't have to specify.
  - You can also use environment variable such as `GCLOUD_PROJECT`.
- `key` (optional)
  - Set your credential file path.
  - Running fluentd on GCP, you can use scope instead of specifying this.
  - You can also use environment variable such as `GCLOUD_KEYFILE`.
- `topic` (required)
  - Set topic name to publish.
- `autocreate_topic` (optional, default: `false`)
  - If set to `true`, specified topic will be created when it doesn't exist.
- `max_messages` (optional, default: `1000`)
  - Publishing messages count per request to Cloud Pub/Sub.
    - See https://cloud.google.com/pubsub/quotas#other_limits
- `max_total_size` (optional, default: `9800000` = `9.8MB`)
  - Publishing messages bytesize per request to Cloud Pub/Sub. This parameter affects only message size. You should specify a little smaller value than quota.
    - See https://cloud.google.com/pubsub/quotas#other_limits
- `max_message_size` (optional, default: `4000000` = `4MB`)
  - Messages exceeding `max_message_size` are not published because Pub/Sub clients cannot receive it.
- `buffer_type`, `buffer_path`, `flush_interval`, `try_flush_interval`
  - These are fluentd buffer configuration. See http://docs.fluentd.org/articles/buffer-plugin-overview
- `format` (optional, default: `json`)
  - Set output format. See http://docs.fluentd.org/articles/out_file#format

### Pull messages

Use `gcloud_pubsub` input plugin.

```
<source>
  @type gcloud_pubsub
  tag example.pull
  project <YOUR PROJECT>
  key <YOUR KEY>
  topic <YOUR TOPIC>
  subscription <YOUR SUBSCRIPTION>
  max_messages 1000
  return_immediately true
  pull_interval 0.5
  pull_threads 2
  format json
  parse_error_action exception
  enable_rpc true
  rpc_bind 0.0.0.0
  rpc_port 24680
</source>
```

- `tag` (required)
  - Set tag of messages.
  - If `tag_key` is specified, `tag` is used as tag when record don't have specified key.
- `tag_key` (optional)
  - Set key to be used as tag.
- `project` (optional)
  - Set your GCP project
  - Running fluentd on GCP, you don't have to specify.
  - You can also use environment variable such as `GCLOUD_PROJECT`.
- `key` (optional)
  - Set your credential file path.
  - Running fluentd on GCP, you can use scope instead of specifying this.
  - You can also use environment variable such as `GCLOUD_KEYFILE`.
- `topic` (required)
  - Set topic name to pull.
- `subscription` (required)
  - Set subscription name to pull.
- `max_messages` (optional, default: `100`)
  - See maxMessages on https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/pull
- `return_immediately` (optional, default: `true`)
  - See returnImmediately on https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/pull
  - If `return_immediately` is `true` or pulling message is stopped by HTTP RPC, this plugin wait `pull_interval` each pull.
- `pull_interval` (optional, default: `5.0`)
  - Pulling messages by intervals of specified seconds.
- `pull_threads` (optional, default: `1`)
  - Set number of threads to pull messages.
- `format` (optional, default: `json`)
  - Set input format. See format section in http://docs.fluentd.org/articles/in_tail
- `parse_error_action` (optional, default: `exception`)
  - Set error type when parsing messages fails.
    - `exception`: Raise exception. Messages are not acknowledged.
    - `warning`: Only logging as warning.
- `enable_rpc` (optional, default: `false`)
  - If `true` is specified, HTTP RPC to stop or start pulling message is enabled.
- `rpc_bind` (optional, default: `0.0.0.0`)
  - Bind IP address for HTTP RPC.
- `rpc_port` (optional, default: `24680`)
  - Port for HTTP RPC.

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request

## TODO

- Add `tag` attribute in output plugin and use `tag` attribute as tag in input plugin.
- Send ack after other output plugin committed (if possible).

## Authors

- [@mdoi](https://github.com/mdoi) : First author
- [@mia-0032](https://github.com/mia-0032) : This version author


CHANGELOG.md
## ChangeLog

### Release 1.0.2 - 2017/09/11

- Bump up google-cloud-pubsub to v0.27.x

### Release 1.0.1 - 2017/09/05

- Bump up google-cloud-pubsub to v0.26.x

### Release 1.0.0 - 2017/06/23

- Fluentd v0.14 ready
  - Fluentd v0.12 is not supported in the later version

### Release 0.4.6 - 2017/05/14

- Output plugin
  - Make messages exceeding configured size not be published because Pub/Sub clients cannot receive it

### Release 0.4.5 - 2017/04/02

- Bump up google-cloud-pubsub to v0.24.x

### Release 0.4.4 - 2017/03/07

- Bump up google-cloud-pubsub to v0.23.x

### Release 0.4.3 - 2017/02/16

- Input plugin
  - Add "status" method to the http rpc api

### Release 0.4.2 - 2017/02/03

- Make retry to get topic/subscription when Pub/Sub API returns 50x code

### Release 0.4.1 - 2017/02/02

- Bump up google-cloud-pubsub to v0.22.x
- Input plugin
  - Rescue 50x errors on acknowledge api

### Release 0.4.0 - 2017/01/21

- Input plugin
  - Add feature to use record key as tag

### Release 0.3.4 - 2017/01/03

- Output plugin
  - Rescue 50x errors
- Input plugin
  - Guard emit to be called with multi-threading
  - Rescue 50x errors
  - Enabled to select whether to raise an exception if message processing failed

### Release 0.3.3 - 2016/12/03

- Input plugin
  - Fix undefined variable error

### Release 0.3.2 - 2016/11/13

- Add plugin param desc
- Input plugin
  - Improve handling to acknowledge messages

### Release 0.3.1 - 2016/11/03

- Output plugin
  - Improve error handling

### Release 0.3.0 - 2016/10/30

- Bump up google-cloud-pubsub to v0.21
- Input plugin
  - Add multithreaded pulling feature

### Release 0.2.0 - 2016/10/15

- Input plugin
  - Add HTTP RPC feature

### Release 0.1.4 - 2016/09/19

- Input plugin
  - `pull_interval` can be specified float value
  - `topic` must be specified

### Release 0.1.3 - 2016/09/17

- Input plugin
  - Fix error handling and add debug logging

### Release 0.1.2 - 2016/09/11

- Output plugin
  - Change default max message size and add debug message

### Release 0.1.1 - 2016/08/27

- Bump up google-cloud-pubsub (gcloud-ruby) to 0.20

### Release 0.1.0 - 2016/08/22

- Use formatter / parser plugin and add format configuration
- Bump up gcloud-ruby to 0.12
- Remove dependency on lightening buffer
- Fix error caused by Pub/Sub quotas


test/plugin/test_in_gcloud_pubsub.rb
require 'net/http'
require 'json'

require_relative "../test_helper"
require "fluent/test/driver/input"

class GcloudPubSubInputTest < Test::Unit::TestCase
  CONFIG = %[
      tag test
      project project-test
      topic topic-test
      subscription subscription-test
      key key-test
      format json
      json_parser yajl
  ]

  DEFAULT_HOST = '127.0.0.1'
  DEFAULT_PORT = 24680

  class DummyInvalidMsgData
    def data
      return 'foo:bar'
    end
  end
  class DummyInvalidMessage
    def message
      DummyInvalidMsgData.new
    end
  end

  def create_driver(conf=CONFIG)
    Fluent::Test::Driver::Input.new(Fluent::Plugin::GcloudPubSubInput).configure(conf)
  end

  def http_get(path)
    http = Net::HTTP.new(DEFAULT_HOST, DEFAULT_PORT)
    req = Net::HTTP::Get.new(path, {'Content-Type' => 'application/x-www-form-urlencoded'})
    http.request(req)
  end

  setup do
    Fluent::Test.setup
  end

  sub_test_case 'configure' do
    test 'all params are configured' do
      d = create_driver(%[
        tag test
        project project-test
        topic topic-test
        subscription subscription-test
        key key-test
        max_messages 1000
        return_immediately true
        pull_interval 2
        pull_threads 3
        format ltsv
        enable_rpc true
        rpc_bind 127.0.0.1
        rpc_port 24681
      ])

      assert_equal('test', d.instance.tag)
      assert_equal('project-test', d.instance.project)
      assert_equal('topic-test', d.instance.topic)
      assert_equal('subscription-test', d.instance.subscription)
      assert_equal('key-test', d.instance.key)
      assert_equal(2.0, d.instance.pull_interval)
      assert_equal(1000, d.instance.max_messages)
      assert_equal(true, d.instance.return_immediately)
      assert_equal(3, d.instance.pull_threads)
      assert_equal('ltsv', d.instance.format)
      assert_equal(true, d.instance.enable_rpc)
      assert_equal('127.0.0.1', d.instance.rpc_bind)
      assert_equal(24681, d.instance.rpc_port)
    end

    test 'default values are configured' do
      d = create_driver
      assert_equal(5.0, d.instance.pull_interval)
      assert_equal(100, d.instance.max_messages)
      assert_equal(true, d.instance.return_immediately)
      assert_equal(1, d.instance.pull_threads)
      assert_equal('json', d.instance.format)
      assert_equal(false, d.instance.enable_rpc)
      assert_equal('0.0.0.0', d.instance.rpc_bind)
      assert_equal(24680, d.instance.rpc_port)
    end
  end

  sub_test_case 'start' do
    setup do
      @topic_mock = mock!
      @pubsub_mock = mock!.topic('topic-test').at_least(1) { @topic_mock }
      stub(Google::Cloud::Pubsub).new { @pubsub_mock }
    end

    test '40x error occurred on connecting to Pub/Sub' do
      @topic_mock.subscription('subscription-test').once do
        raise Google::Cloud::NotFoundError.new('TEST')
      end

      d = create_driver
      assert_raise Google::Cloud::NotFoundError do
        d.run {}
      end
    end

    test '50x error occurred on connecting to Pub/Sub' do
      @topic_mock.subscription('subscription-test').once do
        raise Google::Cloud::UnavailableError.new('TEST')
      end

      d = create_driver
      assert_raise Google::Cloud::UnavailableError do
        d.run {}
      end
    end

    test 'subscription is nil' do
      @topic_mock.subscription('subscription-test').once { nil }

      d = create_driver
      assert_raise Fluent::GcloudPubSub::Error do
        d.run {}
      end
    end
  end

  sub_test_case 'emit' do
    class DummyMsgData
      def data
        return '{"foo": "bar"}'
      end
    end
    class DummyMessage
      def message
        DummyMsgData.new
      end
    end

    class DummyMsgDataWithTagKey
      def initialize(tag)
        @tag = tag
      end
      def data
        return '{"foo": "bar", "test_tag_key": "' + @tag + '"}'
      end
    end
    class DummyMessageWithTagKey
      def initialize(tag)
        @tag = tag
      end
      def message
        DummyMsgDataWithTagKey.new @tag
      end
    end

    setup do
      @subscriber = mock!
      @topic_mock = mock!.subscription('subscription-test') { @subscriber }
      @pubsub_mock = mock!.topic('topic-test') { @topic_mock }
      stub(Google::Cloud::Pubsub).new { @pubsub_mock }
    end

    test 'empty' do
      @subscriber.pull(immediate: true, max: 100).at_least(1) { [] }
      @subscriber.acknowledge.times(0)

      d = create_driver
      d.run(expect_emits: 1, timeout: 3)

      assert_true d.events.empty?
    end

    test 'simple' do
      messages = Array.new(1, DummyMessage.new)
      @subscriber.pull(immediate: true, max: 100).at_least(1) { messages }
      @subscriber.acknowledge(messages).at_least(1)

      d = create_driver
      d.run(expect_emits: 1, timeout: 3)
      emits = d.events

      assert(1 <= emits.length)
      emits.each do |tag, time, record|
        assert_equal("test", tag)
        assert_equal({"foo" => "bar"}, record)
      end
    end

    test 'multithread' do
      messages = Array.new(1, DummyMessage.new)
      @subscriber.pull(immediate: true, max: 100).at_least(2) { messages }
      @subscriber.acknowledge(messages).at_least(2)

      d = create_driver("#{CONFIG}\npull_threads 2")
      d.run(expect_emits: 2, timeout: 1)
      emits = d.events

      assert(2 <= emits.length)
      emits.each do |tag, time, record|
        assert_equal("test", tag)
        assert_equal({"foo" => "bar"}, record)
      end
    end

    test 'with tag_key' do
      messages = [
        DummyMessageWithTagKey.new('tag1'),
        DummyMessageWithTagKey.new('tag2'),
        DummyMessage.new
      ]
      @subscriber.pull(immediate: true, max: 100).at_least(1) { messages }
      @subscriber.acknowledge(messages).at_least(1)

      d = create_driver("#{CONFIG}\ntag_key test_tag_key")
      d.run(expect_emits: 1, timeout: 3)
      emits = d.events

      assert(3 <= emits.length)
      # test tag
      assert_equal("tag1", emits[0][0])
      assert_equal("tag2", emits[1][0])
      assert_equal("test", emits[2][0])
      # test record
      emits.each do |tag, time, record|
        assert_equal({"foo" => "bar"}, record)
      end
    end

    test 'invalid messages with parse_error_action exception ' do
      messages = Array.new(1, DummyInvalidMessage.new)
      @subscriber.pull(immediate: true, max: 100).at_least(1) { messages }
      @subscriber.acknowledge.times(0)

      d = create_driver
      d.run(expect_emits: 1, timeout: 3)
      assert_true d.events.empty?
    end

    test 'invalid messages with parse_error_action warning' do
      messages = Array.new(1, DummyInvalidMessage.new)
      @subscriber.pull(immediate: true, max: 100).at_least(1) { messages }
      @subscriber.acknowledge(messages).at_least(1)

      d = create_driver("#{CONFIG}\nparse_error_action warning")
      d.run(expect_emits: 1, timeout: 3)
      assert_true d.events.empty?
    end

    test 'retry if raised error' do
      class UnknownError < StandardError
      end
      @subscriber.pull(immediate: true, max: 100).at_least(2) { raise UnknownError.new('test') }
      @subscriber.acknowledge.times(0)

      d = create_driver(CONFIG + 'pull_interval 0.5')
      d.run(expect_emits: 1, timeout: 0.8)

      assert_equal(0.5, d.instance.pull_interval)
      assert_true d.events.empty?
    end

    test 'retry if raised RetryableError on pull' do
      @subscriber.pull(immediate: true, max: 100).at_least(2) { raise Google::Cloud::UnavailableError.new('TEST') }
      @subscriber.acknowledge.times(0)

      d = create_driver("#{CONFIG}\npull_interval 0.5")
      d.run(expect_emits: 1, timeout: 0.8)

      assert_equal(0.5, d.instance.pull_interval)
      assert_true d.events.empty?
    end

    test 'retry if raised RetryableError on acknowledge' do
      messages = Array.new(1, DummyMessage.new)
      @subscriber.pull(immediate: true, max: 100).at_least(2) { messages }
      @subscriber.acknowledge(messages).at_least(2) { raise Google::Cloud::UnavailableError.new('TEST') }

      d = create_driver("#{CONFIG}\npull_interval 0.5")
      d.run(expect_emits: 2, timeout: 3)
      emits = d.events

      # not acknowledged, but already emitted to engine.
      assert(2 <= emits.length)
      emits.each do |tag, time, record|
        assert_equal("test", tag)
        assert_equal({"foo" => "bar"}, record)
      end
    end

    test 'stop by http rpc' do
      messages = Array.new(1, DummyMessage.new)
      @subscriber.pull(immediate: true, max: 100).once { messages }
      @subscriber.acknowledge(messages).once

      d = create_driver("#{CONFIG}\npull_interval 1.0\nenable_rpc true")
      assert_equal(false, d.instance.instance_variable_get(:@stop_pull))

      d.run {
        http_get('/api/in_gcloud_pubsub/pull/stop')
        sleep 0.75
        # d.run sleeps 0.5 sec
      }
      emits = d.events

      assert_equal(1, emits.length)
      assert_true d.instance.instance_variable_get(:@stop_pull)

      emits.each do |tag, time, record|
        assert_equal("test", tag)
        assert_equal({"foo" => "bar"}, record)
      end
    end

    test 'start by http rpc' do
      messages = Array.new(1, DummyMessage.new)
      @subscriber.pull(immediate: true, max: 100).at_least(1) { messages }
      @subscriber.acknowledge(messages).at_least(1)

      d = create_driver("#{CONFIG}\npull_interval 1.0\nenable_rpc true")
      d.instance.stop_pull
      assert_equal(true, d.instance.instance_variable_get(:@stop_pull))

      d.run(expect_emits: 1, timeout: 3) {
        http_get('/api/in_gcloud_pubsub/pull/start')
        sleep 0.75
        # d.run sleeps 0.5 sec
      }
      emits = d.events

      assert_equal(true, emits.length > 0)
      assert_false d.instance.instance_variable_get(:@stop_pull)

      emits.each do |tag, time, record|
        assert_equal("test", tag)
        assert_equal({"foo" => "bar"}, record)
      end
    end

    test 'get status by http rpc when started' do
      d = create_driver("#{CONFIG}\npull_interval 1.0\nenable_rpc true")
      assert_false d.instance.instance_variable_get(:@stop_pull)

      d.run {
        res = http_get('/api/in_gcloud_pubsub/pull/status')
        assert_equal({"ok" => true, "status" => "started"}, JSON.parse(res.body))
      }
    end

    test 'get status by http rpc when stopped' do
      d = create_driver("#{CONFIG}\npull_interval 1.0\nenable_rpc true")
      d.instance.stop_pull
      assert_true d.instance.instance_variable_get(:@stop_pull)

      d.run {
        res = http_get('/api/in_gcloud_pubsub/pull/status')
        assert_equal({"ok" => true, "status" => "stopped"}, JSON.parse(res.body))
      }
    end
  end
end


test/plugin/test_out_gcloud_pubsub.rb
# coding: utf-8
require_relative "../test_helper"
require "fluent/test/driver/output"
require "fluent/test/helpers"

class GcloudPubSubOutputTest < Test::Unit::TestCase
  include Fluent::Test::Helpers

  CONFIG = %[
    project project-test
    topic topic-test
    key key-test
  ]

  ReRaisedError = Class.new(RuntimeError)

  def create_driver(conf = CONFIG)
    Fluent::Test::Driver::Output.new(Fluent::Plugin::GcloudPubSubOutput).configure(conf)
  end

  setup do
    Fluent::Test.setup
  end

  sub_test_case 'configure' do
    test 'default values are configured' do
      d = create_driver(%[
        project project-test
        topic topic-test
        key key-test
      ])

      assert_equal('project-test', d.instance.project)
      assert_equal('topic-test', d.instance.topic)
      assert_equal('key-test', d.instance.key)
      assert_equal(false, d.instance.autocreate_topic)
      assert_equal(1000, d.instance.max_messages)
      assert_equal(9800000, d.instance.max_total_size)
      assert_equal(4000000, d.instance.max_message_size)
    end

    test '"topic" must be specified' do
      assert_raises Fluent::ConfigError do
        create_driver(%[
          project project-test
          key key-test
        ])
      end
    end

    test '"autocreate_topic" can be specified' do
      d = create_driver(%[
        project project-test
        topic topic-test
        key key-test
        autocreate_topic true
      ])

      assert_equal(true, d.instance.autocreate_topic)
    end
  end

  sub_test_case 'topic' do
    setup do
      @publisher = mock!
      @pubsub_mock = mock!
      stub(Google::Cloud::Pubsub).new { @pubsub_mock }
    end

    test '"autocreate_topic" is enabled' do
      d = create_driver(%[
        project project-test
        topic topic-test
        key key-test
        autocreate_topic true
      ])

      @pubsub_mock.topic("topic-test").once { nil }
      @pubsub_mock.create_topic("topic-test").once { @publisher }
      d.run
    end

    test '40x error occurred on connecting to Pub/Sub' do
      d = create_driver

      @pubsub_mock.topic('topic-test').once do
        raise Google::Cloud::NotFoundError.new('TEST')
      end

      assert_raise Google::Cloud::NotFoundError do
        d.run {}
      end
    end

    test '50x error occurred on connecting to Pub/Sub' do
      d = create_driver

      @pubsub_mock.topic('topic-test').once do
        raise Google::Cloud::UnavailableError.new('TEST')
      end

      assert_raise Google::Cloud::UnavailableError do
        d.run {}
      end
    end

    test 'topic is nil' do
      d = create_driver

      @pubsub_mock.topic('topic-test').once { nil }

      assert_raise Fluent::GcloudPubSub::Error do
        d.run {}
      end
    end
  end

  sub_test_case 'publish' do
    setup do
      @publisher = mock!
      @pubsub_mock = mock!.topic(anything) { @publisher }
      stub(Google::Cloud::Pubsub).new { @pubsub_mock }
    end

    setup do
      @time = event_time('2016-07-09 11:12:13 UTC')
    end

    test 'messages are divided into "max_messages"' do
      d = create_driver
      @publisher.publish.times(2)
      d.run(default_tag: "test") do
        # max_messages is default 1000
        1001.times do |i|
          d.feed(@time, {"a" => i})
        end
      end
    end

    test 'messages are divided into "max_total_size"' do
      d = create_driver(%[
        project project-test
        topic topic-test
        key key-test
        max_messages 100000
        max_total_size 1000
      ])

      @publisher.publish.times(2)
      d.run(default_tag: "test") do
        # 400 * 4 / max_total_size = twice
        4.times do
          d.feed(@time, {"a" => "a" * 400})
        end
      end
    end

    test 'messages exceeding "max_message_size" are not published' do
      d = create_driver(%[
        project project-test
        topic topic-test
        key key-test
        max_message_size 1000
      ])

      @publisher.publish.times(0)
      d.run(default_tag: "test") do
        d.feed(@time, {"a" => "a" * 1000})
      end
    end

    test 'accept "ASCII-8BIT" encoded multibyte strings' do
      # on fluentd v0.14, all strings treated as "ASCII-8BIT" except specified encoding.
      d = create_driver
      @publisher.publish.once
      d.run(default_tag: "test") do
        d.feed(@time, {"a" => "あああ".force_encoding("ASCII-8BIT")})
      end
    end

    test 'reraise unexpected errors' do
      d = create_driver
      @publisher.publish.once { raise ReRaisedError }
      assert_raises ReRaisedError do
        d.run(default_tag: "test") do
          d.feed([{'a' => 1, 'b' => 2}])
        end
      end
    end

    test 'reraise RetryableError' do
      d = create_driver
      @publisher.publish.once { raise Google::Cloud::UnavailableError.new('TEST') }
      assert_raises Fluent::GcloudPubSub::RetryableError do
        d.run(default_tag: "test") do
          d.feed([{'a' => 1, 'b' => 2}])
        end
      end
    end
  end
end


lib/fluent/plugin/out_gcloud_pubsub.rb
require 'fluent/plugin/output'

require 'fluent/plugin/gcloud_pubsub/client'

module Fluent::Plugin
  class GcloudPubSubOutput < Output
    Fluent::Plugin.register_output('gcloud_pubsub', self)

    helpers :compat_parameters, :formatter

    DEFAULT_BUFFER_TYPE = "memory"
    DEFAULT_FORMATTER_TYPE = "json"

    desc 'Set your GCP project.'
    config_param :project,            :string,  :default => nil
    desc 'Set your credential file path.'
    config_param :key,                :string,  :default => nil
    desc 'Set topic name to publish.'
    config_param :topic,              :string
    desc "If set to `true`, specified topic will be created when it doesn't exist."
    config_param :autocreate_topic,   :bool,    :default => false
    desc 'Publishing messages count per request to Cloud Pub/Sub.'
    config_param :max_messages,       :integer, :default => 1000
    desc 'Publishing messages bytesize per request to Cloud Pub/Sub.'
    config_param :max_total_size,     :integer, :default => 9800000  # 9.8MB
    desc 'Limit bytesize per message.'
    config_param :max_message_size,   :integer, :default => 4000000  # 4MB
    desc 'Set output format.'
    config_param :format,             :string,  :default => 'json'

    config_section :buffer do
      config_set_default :@type, DEFAULT_BUFFER_TYPE
    end

    config_section :format do
      config_set_default :@type, DEFAULT_FORMATTER_TYPE
    end

    def configure(conf)
      compat_parameters_convert(conf, :buffer, :formatter)
      super
      placeholder_validate!(:topic, @topic)
      @formatter = formatter_create
    end

    def start
      super
      @publisher = Fluent::GcloudPubSub::Publisher.new @project, @key, @autocreate_topic
    end

    def format(tag, time, record)
      @formatter.format(tag, time, record).to_msgpack
    end

    def formatted_to_msgpack_binary?
      true
    end

    def multi_workers_ready?
      true
    end

    def write(chunk)
      topic = extract_placeholders(@topic, chunk.metadata)

      messages = []
      size = 0

      chunk.msgpack_each do |msg|
        if msg.bytesize > @max_message_size
          log.warn 'Drop a message because its size exceeds `max_message_size`', size: msg.bytesize
          next
        end
        if messages.length + 1 > @max_messages || size + msg.bytesize > @max_total_size
          publish(topic, messages)
          messages = []
          size = 0
        end
        messages << msg
        size += msg.bytesize
      end

      if messages.length > 0
        publish(topic, messages)
      end
    rescue Fluent::GcloudPubSub::RetryableError => ex
      log.warn "Retryable error occurs. Fluentd will retry.", error_message: ex.to_s, error_class: ex.class.to_s
      raise ex
    rescue => ex
      log.error "unexpected error", error_message: ex.to_s, error_class: ex.class.to_s
      log.error_backtrace
      raise ex
    end

    private

    def publish(topic, messages)
      log.debug "send message topic:#{topic} length:#{messages.length} size:#{messages.map(&:bytesize).inject(:+)}"
      @publisher.publish(topic, messages)
    end
  end
end


lib/fluent/plugin/in_gcloud_pubsub.rb
require 'json'
require 'webrick'

require 'fluent/plugin/input'
require 'fluent/plugin/parser'

require 'fluent/plugin/gcloud_pubsub/client'

module Fluent::Plugin
  class GcloudPubSubInput < Input
    Fluent::Plugin.register_input('gcloud_pubsub', self)

    helpers :compat_parameters, :parser, :thread

    DEFAULT_PARSER_TYPE = 'json'

    class FailedParseError < StandardError
    end

    desc 'Set tag of messages.'
    config_param :tag,                :string
    desc 'Set key to be used as tag.'
    config_param :tag_key,            :string,  default: nil
    desc 'Set your GCP project.'
    config_param :project,            :string,  default: nil
    desc 'Set your credential file path.'
    config_param :key,                :string,  default: nil
    desc 'Set topic name to pull.'
    config_param :topic,              :string
    desc 'Set subscription name to pull.'
    config_param :subscription,       :string
    desc 'Pulling messages by intervals of specified seconds.'
    config_param :pull_interval,      :float,   default: 5.0
    desc 'Max messages pulling at once.'
    config_param :max_messages,       :integer, default: 100
    desc 'Setting `true`, keepalive connection to wait for new messages.'
    config_param :return_immediately, :bool,    default: true
    desc 'Set number of threads to pull messages.'
    config_param :pull_threads,       :integer, default: 1
    desc 'Set input format.'
    config_param :format,             :string,  default: DEFAULT_PARSER_TYPE
    desc 'Set error type when parsing messages fails.'
    config_param :parse_error_action, :enum,    default: :exception, list: [:exception, :warning]
    # for HTTP RPC
    desc 'If `true` is specified, HTTP RPC to stop or start pulling message is enabled.'
    config_param :enable_rpc,         :bool,    default: false
    desc 'Bind IP address for HTTP RPC.'
    config_param :rpc_bind,           :string,  default: '0.0.0.0'
    desc 'Port for HTTP RPC.'
    config_param :rpc_port,           :integer, default: 24680

    config_section :parse do
      config_set_default :@type, DEFAULT_PARSER_TYPE
    end

    class RPCServlet < WEBrick::HTTPServlet::AbstractServlet
      class Error < StandardError; end

      def initialize(server, plugin)
        super
        @plugin = plugin
      end

      def do_GET(req, res)
        begin
          code, header, body = process(req, res)
        rescue
          code, header, body = render_json(500, {
              'ok' => false,
              'message' => 'Internal Server Error',
              'error' => "#{$!}",
              'backtrace'=> $!.backtrace
          })
        end

        res.status = code
        header.each_pair {|k,v|
          res[k] = v
        }
        res.body = body
      end

      def render_json(code, obj)
        [code, {'Content-Type' => 'application/json'}, obj.to_json]
      end

      def process(req, res)
        ret = {'ok' => true}
        case req.path_info
        when '/stop'
          @plugin.stop_pull
        when '/start'
          @plugin.start_pull
        when '/status'
          ret['status'] = @plugin.status_of_pull
        else
          raise Error.new "Invalid path_info: #{req.path_info}"
        end
        render_json(200, ret)
      end
    end

    def configure(conf)
      compat_parameters_convert(conf, :parser)
      super
      @rpc_srv = nil
      @rpc_thread = nil
      @stop_pull = false

      @extract_tag = if @tag_key.nil?
                       method(:static_tag)
                     else
                       method(:dynamic_tag)
                     end

      @parser = parser_create
    end

    def start
      super
      start_rpc if @enable_rpc

      @subscriber = Fluent::GcloudPubSub::Subscriber.new @project, @key, @topic, @subscription
      log.debug "connected subscription:#{@subscription} in project #{@project}"

      @emit_guard = Mutex.new
      @stop_subscribing = false
      @subscribe_threads = []
      @pull_threads.times do |idx|
        @subscribe_threads.push thread_create("in_gcloud_pubsub_subscribe_#{idx}".to_sym, &method(:subscribe))
      end
    end

    def shutdown
      if @rpc_srv
        @rpc_srv.shutdown
        @rpc_srv = nil
      end
      if @rpc_thread
        @rpc_thread = nil
      end
      @stop_subscribing = true
      @subscribe_threads.each(&:join)
      super
    end

    def stop_pull
      @stop_pull = true
      log.info "stop pull from subscription:#{@subscription}"
    end

    def start_pull
      @stop_pull = false
      log.info "start pull from subscription:#{@subscription}"
    end

    def status_of_pull
      @stop_pull ? 'stopped' : 'started'
    end

    private

    def static_tag(record)
      @tag
    end

    def dynamic_tag(record)
      record.delete(@tag_key) || @tag
    end

    def start_rpc
      log.info "listening http rpc server on http://#{@rpc_bind}:#{@rpc_port}/"
      @rpc_srv = WEBrick::HTTPServer.new(
        {
          BindAddress: @rpc_bind,
          Port: @rpc_port,
          Logger: WEBrick::Log.new(STDERR, WEBrick::Log::FATAL),
          AccessLog: []
        }
      )
      @rpc_srv.mount('/api/in_gcloud_pubsub/pull/', RPCServlet, self)
      @rpc_thread = thread_create(:in_gcloud_pubsub_rpc_thread){
        @rpc_srv.start
      }
    end

    def subscribe
      until @stop_subscribing
        _subscribe unless @stop_pull

        if @return_immediately || @stop_pull
          sleep @pull_interval
        end
      end
    rescue => ex
      log.error "unexpected error", error_message: ex.to_s, error_class: ex.class.to_s
      log.error_backtrace ex.backtrace
    end

    def _subscribe
      messages = @subscriber.pull @return_immediately, @max_messages
      if messages.length == 0
        log.debug "no messages are pulled"
        return
      end

      process messages
      @subscriber.acknowledge messages

      log.debug "#{messages.length} message(s) processed"
    rescue Fluent::GcloudPubSub::RetryableError => ex
      log.warn "Retryable error occurs. Fluentd will retry.", error_message: ex.to_s, error_class: ex.class.to_s
    rescue => ex
      log.error "unexpected error", error_message: ex.to_s, error_class: ex.class.to_s
      log.error_backtrace ex.backtrace
    end

    def process(messages)
      event_streams = Hash.new do |hsh, key|
        hsh[key] = Fluent::MultiEventStream.new
      end

      messages.each do |m|
        line = m.message.data.chomp
        @parser.parse(line) do |time, record|
          if time && record
            event_streams[@extract_tag.call(record)].add(time, record)
          else
            case @parse_error_action
            when :exception
              raise FailedParseError.new "pattern not match: #{line.inspect}"
            else
              log.warn 'pattern not match', record: line.inspect
            end
          end
        end
      end

      event_streams.each do |tag, es|
        # There are some output plugins not to supposed to be called with multi-threading.
        # Maybe remove in the future.
        @emit_guard.synchronize do
          router.emit_stream(tag, es)
        end
      end
    end
  end
end
